{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cores = '4'\n",
    "strains = 'https://evocellnet.github.io/ecoref/data/strains.tsv'\n",
    "phenotypes = 'https://evocellnet.github.io/ecoref/data/phenotypic_data.tsv'\n",
    "pathogenicity = '../data/phenotypes/phenotypes.tsv'\n",
    "gdir = '../data/genomes'\n",
    "filtered = '../out/associations/summary_cont_lmm_kmer.tsv'\n",
    "rtab = '../out/roary/gene_presence_absence.Rtab'\n",
    "hpi = '../data/hpi.tsv'\n",
    "binary = '../out/associations/annotated_binary_cont_lmm_kmer.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# other genes in the vicinity of HPI\n",
    "otherhpi = {'group_1493', 'group_526', 'group_181'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cores = int(cores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotting imports\n",
    "%matplotlib inline\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from adjustText import adjust_text\n",
    "\n",
    "sns.set_style('white')\n",
    "\n",
    "plt.rc('font', size=11)\n",
    "plt.rc('xtick', labelsize=11)\n",
    "plt.rc('ytick', labelsize=11)\n",
    "plt.rc('axes', labelsize=12, titlesize=12)\n",
    "plt.rc('legend', fontsize=11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "from sklearn import impute\n",
    "from sklearn import metrics\n",
    "from sklearn import model_selection\n",
    "from sklearn.ensemble import RandomForestClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rhp = [x.rstrip().split()[0]\n",
    "       for x in open(hpi)]\n",
    "hp = [x.rstrip().split()[1]\n",
    "      for x in open(hpi)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def match(x, y, c,\n",
    "          xcolumn='s-scores',\n",
    "          ycolumn='killed'):\n",
    "    j = x.loc[c, xcolumn]\n",
    "    k = y.loc[j.index, ycolumn]\n",
    "    idx = j.index.intersection(k.index)\n",
    "    if len(idx) == 0:\n",
    "        return np.nan, np.nan\n",
    "    j = j.loc[idx]\n",
    "    k = k.loc[idx]\n",
    "    return j, k\n",
    "\n",
    "def correlate(x, y, c,\n",
    "              method='pearson',\n",
    "              xcolumn='s-scores',\n",
    "              ycolumn='killed'):\n",
    "    j, k = match(x, y, c, xcolumn, ycolumn)\n",
    "    if method == 'pearson':\n",
    "        return stats.pearsonr(j, k)\n",
    "    else:\n",
    "        return stats.spearmanr(j, k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rndmfrst(x, y, xcolumn='s-scores', ycolumn='phenotype', test_size=0.33):\n",
    "    c = x[xcolumn].unstack()\n",
    "    c = pd.DataFrame(impute.SimpleImputer().fit_transform(c),\n",
    "                     index=c.index,\n",
    "                     columns=c.columns).T\n",
    "    c['mouse'] = y.loc[c.index, ycolumn]\n",
    "    \n",
    "    df = c[sorted(set(c.columns) - {'mouse'})]\n",
    "    target = c['mouse']\n",
    "    \n",
    "    cv = model_selection.StratifiedShuffleSplit(n_splits=1,\n",
    "                                                test_size=test_size,\n",
    "                                                random_state=np.random.RandomState(42))\n",
    "    train_idx, test_idx = next(cv.split(df, target))\n",
    "    \n",
    "    df_train, target_train = df.iloc[train_idx], target[train_idx]\n",
    "    df_test, target_test = df.iloc[test_idx], target[test_idx]\n",
    "    \n",
    "    param_grid = {'n_estimators': [int(x) for x in np.logspace(1, 3, 6)],\n",
    "                  'max_features': range(int(np.sqrt(df.shape[1])), int(df.shape[1] *0.7), 5)}\n",
    "\n",
    "    clf = RandomForestClassifier(random_state=np.random.RandomState(42))\n",
    "    grid_search = model_selection.GridSearchCV(clf,\n",
    "                                               param_grid=param_grid,\n",
    "                                               cv=model_selection.StratifiedShuffleSplit(n_splits=10,\n",
    "                                                              test_size=0.33,\n",
    "                                                              random_state=np.random.RandomState(42)),\n",
    "                                               scoring='f1',\n",
    "                                               n_jobs=cores)\n",
    "    grid_search.fit(df_train, target_train)\n",
    "    \n",
    "    clf = RandomForestClassifier(n_estimators=grid_search.best_params_['n_estimators'],\n",
    "                                 max_features=grid_search.best_params_['max_features'],\n",
    "                                 random_state=np.random.RandomState(42))\n",
    "    clf.fit(df_train, target_train)\n",
    "    predict_test = pd.Series(clf.predict(df_test),\n",
    "                             index=target_test.index,\n",
    "                             name='prediction')\n",
    "    combined = target_test.to_frame().join(predict_test.to_frame())\n",
    "    print(metrics.f1_score(combined['mouse'],\n",
    "                           combined['prediction']),\n",
    "          grid_search.best_score_)\n",
    "    \n",
    "    return clf, combined, df_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = pd.read_table(filtered,\n",
    "                  index_col=0)\n",
    "f = f[f['specific_hits'] > 0]\n",
    "r = pd.read_table(rtab, index_col=0)\n",
    "r = r.loc[f[f['maxp'] > 15].index].copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hpi = r.loc[rhp].sum()\n",
    "hpi.name = 'hpi'\n",
    "hpi = hpi.to_frame()\n",
    "hpi['present'] = [1 if x >= 10 \n",
    "                  else 0\n",
    "                  for x in hpi['hpi'].values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nhpi = pd.read_csv(binary, sep='\\t', index_col=0)\n",
    "nhpi = nhpi.loc[r.index.difference(rhp).difference(otherhpi)].sum().to_frame().rename(columns={0: 'nhpi'})\n",
    "nhpi['present'] = [1 if x > 3\n",
    "                   else 0\n",
    "                   for x in nhpi['nhpi'].values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exclude = {'ECOR-03', 'ECOR-05', 'ECOR-06', 'ECOR-07',\n",
    "           'ECOR-08', 'ECOR-22', 'ECOR-23', 'ECOR-30',\n",
    "           'ECOR-31', 'ECOR-32', 'ECOR-33', 'ECOR-35',\n",
    "           'ECOR-36', 'ECOR-37', 'ECOR-43', 'ECOR-44',\n",
    "           'ECOR-45', 'ECOR-47', 'ECOR-48', 'ECOR-49',\n",
    "           'ECOR-51', 'ECOR-55', 'ECOR-56', 'ECOR-61',\n",
    "           'ECOR-64', 'ECOR-65', 'ECOR-66', 'ECOR-68',\n",
    "           'ECOR-71', 'IAI29', 'IAI64', 'NILS80'}\n",
    "genomes = {x.split('.')[0]\n",
    "           for x in os.listdir(gdir)\n",
    "           if x.endswith('.fasta')} - exclude"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = pd.read_table(strains, index_col=1)\n",
    "s.index = [x.replace(' ', '')\n",
    "           if x.startswith('NILS')\n",
    "           else x\n",
    "           for x in s.index]\n",
    "s = s.loc[[k for k,v in s.T.iteritems()\n",
    "           if k in genomes\n",
    "           and v.values[0] != 'NT12008']][['Strain Identifier', 'Full Strain Name After Genome Analysis']].dropna()\n",
    "s = s['Strain Identifier']\n",
    "s = s.reset_index().set_index(\n",
    "    'Strain Identifier'\n",
    "    ).drop('NT12008').reset_index().set_index(\n",
    "    'index')['Strain Identifier']\n",
    "d = {v: k\n",
    "     for k, v in s.iteritems()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = pd.read_table(phenotypes).set_index('strain')\n",
    "p.index = [d.get(x, x)\n",
    "           for x in p.index]\n",
    "p.index.name = 'strain'\n",
    "p = p.reset_index().set_index(['strain', 'condition'])\n",
    "p = p.loc[s.index]\n",
    "p = p.reset_index().set_index(['condition', 'strain'])\n",
    "p = p.sort_index()\n",
    "p['signed-qvalue'] = [np.sign(x) * -np.log10(y)\n",
    "                      for x, y, z in p.values]\n",
    "# idx = p.groupby('condition')['growth-defect-phenotype'\n",
    "#         ].sum()[p.groupby('condition')[\n",
    "#       'growth-defect-phenotype'].sum() > 0].index\n",
    "# p = p.loc[idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "k = pd.read_table(pathogenicity,\n",
    "                  index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = []\n",
    "for c in {x[0] for x in p.index}:\n",
    "    r, pval = correlate(p, k, c)\n",
    "    res.append((c, 'pearson', 's-score', r, pval))\n",
    "    r, pval = correlate(p, k, c, method='spearman')\n",
    "    res.append((c, 'spearman', 's-score', r, pval))\n",
    "    r, pval = correlate(p, k, c, xcolumn='signed-qvalue')\n",
    "    res.append((c, 'pearson', 'signed-qvalue', r, pval))\n",
    "    r, pval = correlate(p, k, c, method='spearman', xcolumn='signed-qvalue')\n",
    "    res.append((c, 'spearman', 'signed-qvalue', r, pval))\n",
    "    r, pval = correlate(p, hpi, c, ycolumn='hpi')\n",
    "    res.append((c, 'pearson', 's-score.hpi', r, pval))\n",
    "    r, pval = correlate(p, hpi, c, method='spearman', ycolumn='hpi')\n",
    "    res.append((c, 'spearman', 's-score.hpi', r, pval))\n",
    "    r, pval = correlate(p, hpi, c, xcolumn='signed-qvalue', ycolumn='hpi')\n",
    "    res.append((c, 'pearson', 'signed-qvalue.hpi', r, pval))\n",
    "    r, pval = correlate(p, hpi, c, method='spearman', xcolumn='signed-qvalue', ycolumn='hpi')\n",
    "    res.append((c, 'spearman', 'signed-qvalue.hpi', r, pval))\n",
    "    r, pval = correlate(p, nhpi, c, ycolumn='nhpi')\n",
    "    res.append((c, 'pearson', 's-score.nhpi', r, pval))\n",
    "    r, pval = correlate(p, nhpi, c, method='spearman', ycolumn='nhpi')\n",
    "    res.append((c, 'spearman', 's-score.nhpi', r, pval))\n",
    "    r, pval = correlate(p, nhpi, c, xcolumn='signed-qvalue', ycolumn='nhpi')\n",
    "    res.append((c, 'pearson', 'signed-qvalue.nhpi', r, pval))\n",
    "    r, pval = correlate(p, nhpi, c, method='spearman', xcolumn='signed-qvalue', ycolumn='nhpi')\n",
    "    res.append((c, 'spearman', 'signed-qvalue.nhpi', r, pval))\n",
    "r = pd.DataFrame(res,\n",
    "                 columns=['condition',\n",
    "                          'method',\n",
    "                          'column',\n",
    "                          'r',\n",
    "                          'p'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "r.to_csv('correlations.tsv', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd = {'NACL.RT': 'NaCl 400$nM$',\n",
    "      'PH5.RT': 'pH5',\n",
    "      'PH9.BILESALTS': 'pH9 + Bile salts',\n",
    "      'PH9C.UV': 'pH9 + UV light',\n",
    "      'PVP.0P04': 'PVP 0.04$\\mu g/ml$',\n",
    "      'A22.CHIR090': 'A22 + CHIR-090',\n",
    "      'THIOLUTIN.10': r'Thiolutin 10$\\mu g/ml$',\n",
    "      'OXACILLIN.0P5UM': r'Oxacillin 0.05$\\mu M$',\n",
    "      'CIPROFLOXACIN.0P001UM': 'Ciprofloxacin 0.001$\\mu M$',\n",
    "      'RIFAMPICIN.1UM': 'Rifampicin 1$\\mu M$',\n",
    "      'RIFAMPICIN.2UM': 'Rifampicin 2$\\mu M$',\n",
    "      'INDOLE.TOBRAMYCIN': 'Indole + Tobramycin',\n",
    "      'TEICOPLANIN.10': 'Teicoplanin 10$\\mu g/ml$',\n",
    "      'INDOLE.CEFSULODIN': 'Indole + Cefsulodin',\n",
    "      'CERULENIN.1UM': 'Cerulenin 1$\\mu M$',\n",
    "      'COLICINM.0P4': 'Colicin-M 0.4$\\mu g/ml$',\n",
    "      'PENTAACETICACID.15': 'Pentetic acid 15$\\mu g/ml$',\n",
    "      'AMOXICILLIN.0P5UM': 'Amoxicillin 0.5$\\mu M$',\n",
    "      'PYOCYANIN.0P2UM': 'Pyocyanin 0.2$\\mu M$',\n",
    "      'MEROPENEM.0P03': 'Meropenem 0.03$\\mu g/ml$',\n",
    "      'PARAQUAT.10UM': 'Paraquat 10$\\mu M$',\n",
    "      'DM.GLUCOSE25MM': 'Glucose 25$mM$',\n",
    "      'PIERICIDIN.20': 'Piericidin 20$\\mu g/ml$',\n",
    "      'TETRACYCLINE.0P5UM': 'Tetracycline 0.5$\\mu M$',\n",
    "      'TETRACYCLINE.0P25UM': 'Tetracycline 0.25$\\mu M$',\n",
    "      'INDOLE.40C': 'Indole 40$^\\circ C$',\n",
    "      'SULFAMONOMETHOXINE.RIFAMPICIN': 'Sulfamonomethoxine + Rifampicin',\n",
    "      'PH9.NITRATEK': 'pH9 + Nitratek',\n",
    "      'PH5.FUMARATE': 'pH5 + Fumarate',\n",
    "      'FUMARATE.40MM': 'Fumarate 40$mM$',\n",
    "      'DM.MELIBIOSE': 'Melibiose 10$mM$',\n",
    "      'CHLORHEXIDINE.0P8': 'Chlorhexidine 0.8 $\\mu ug/ml$',\n",
    "      'SPECTINOMYCIN.NITROFURANTOIN': 'Spectinomycin + Nitrofurantoin',\n",
    "      'A22.0P5UM': 'A22 0.5$\\mu M$',\n",
    "      'TOBRAMYCIN.0P0125UM': 'Tobramycin 0.0125$\\mu M$',\n",
    "      'SORBITOL.500MM': 'Sorbitol 500$mM$'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(4, 4))\n",
    "\n",
    "r1 = r[(r['method'] == 'pearson') &\n",
    "       (r['column'] == 's-score')].copy(deep=True)\n",
    "r1 = r1.set_index('condition')\n",
    "\n",
    "plt.scatter(r1['r'],\n",
    "            -np.log10(r1['p']),\n",
    "            color='k',\n",
    "            marker='.',\n",
    "            rasterized=True)\n",
    "r2 = r1[-np.log10(r1['p']) > 4]\n",
    "plt.scatter(r2['r'],\n",
    "            -np.log10(r2['p']),\n",
    "            color=sns.xkcd_rgb['pale red'],\n",
    "            marker='o')\n",
    "r2 = r1[(-np.log10(r1['p']) > 2) &\n",
    "        (-np.log10(r1['p']) < 4)]\n",
    "plt.scatter(r2['r'],\n",
    "            -np.log10(r2['p']),\n",
    "            color=sns.xkcd_rgb['pale red'],\n",
    "            marker='.')\n",
    "\n",
    "texts = []\n",
    "for t, v in r1[-np.log10(r1['p']) >= 4].iterrows():\n",
    "    texts.append(plt.text(v['r'],\n",
    "                          -np.log10(v['p']),\n",
    "                          cd.get(t, t),\n",
    "                          ha='center',\n",
    "                          va='center',\n",
    "                          fontdict={'size': 9}))\n",
    "# adjust_text(texts,\n",
    "#             arrowprops=dict(arrowstyle='-', color='k'),\n",
    "#             force_points=1)\n",
    "\n",
    "plt.ylim(-0.5, 9)\n",
    "\n",
    "plt.xlabel('Pearson\\'s $r$')\n",
    "plt.ylabel('$-log_{10}(pvalue)$')\n",
    "\n",
    "plt.axvline(0,\n",
    "            color='grey',\n",
    "            ls='dashed')\n",
    "plt.axhline(2,\n",
    "            color='grey',\n",
    "            ls='dashed')\n",
    "plt.axhline(4,\n",
    "            color='grey',\n",
    "            ls='dashed')\n",
    "\n",
    "plt.savefig('volcano_1.png',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True)\n",
    "plt.savefig('volcano_1.svg',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(4, 4))\n",
    "\n",
    "r1 = r[(r['method'] == 'pearson') &\n",
    "       (r['column'] == 's-score.hpi')].copy(deep=True)\n",
    "r1 = r1.set_index('condition')\n",
    "\n",
    "plt.scatter(r1['r'],\n",
    "            -np.log10(r1['p']),\n",
    "            color='k',\n",
    "            marker='.',\n",
    "            rasterized=True)\n",
    "r2 = r1[-np.log10(r1['p']) > 4]\n",
    "plt.scatter(r2['r'],\n",
    "            -np.log10(r2['p']),\n",
    "            color=sns.xkcd_rgb['pale red'],\n",
    "            marker='o')\n",
    "r2 = r1[(-np.log10(r1['p']) > 2) &\n",
    "        (-np.log10(r1['p']) < 4)]\n",
    "plt.scatter(r2['r'],\n",
    "            -np.log10(r2['p']),\n",
    "            color=sns.xkcd_rgb['pale red'],\n",
    "            marker='.')\n",
    "\n",
    "texts = []\n",
    "for t, v in r1[-np.log10(r1['p']) >= 4].iterrows():\n",
    "    texts.append(plt.text(v['r'],\n",
    "                          -np.log10(v['p']),\n",
    "                          cd.get(t, t),\n",
    "                          ha='center',\n",
    "                          va='center',\n",
    "                          fontdict={'size': 9}))\n",
    "# adjust_text(texts,\n",
    "#             arrowprops=dict(arrowstyle='-', color='k'),\n",
    "#             force_points=1)\n",
    "\n",
    "plt.ylim(-0.5, 9)\n",
    "\n",
    "plt.xlabel('Pearson\\'s $r$')\n",
    "plt.ylabel('$-log_{10}(pvalue)$')\n",
    "\n",
    "plt.axvline(0,\n",
    "            color='grey',\n",
    "            ls='dashed')\n",
    "plt.axhline(2,\n",
    "            color='grey',\n",
    "            ls='dashed')\n",
    "plt.axhline(4,\n",
    "            color='grey',\n",
    "            ls='dashed')\n",
    "\n",
    "plt.savefig('volcano_2.png',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True)\n",
    "plt.savefig('volcano_2.svg',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(4, 4))\n",
    "\n",
    "r1 = r[(r['method'] == 'pearson') &\n",
    "       (r['column'] == 's-score.nhpi')].copy(deep=True)\n",
    "r1 = r1.set_index('condition')\n",
    "\n",
    "plt.scatter(r1['r'],\n",
    "            -np.log10(r1['p']),\n",
    "            color='k',\n",
    "            marker='.',\n",
    "            rasterized=True)\n",
    "r2 = r1[-np.log10(r1['p']) > 4]\n",
    "plt.scatter(r2['r'],\n",
    "            -np.log10(r2['p']),\n",
    "            color=sns.xkcd_rgb['pale red'],\n",
    "            marker='o')\n",
    "r2 = r1[(-np.log10(r1['p']) > 2) &\n",
    "        (-np.log10(r1['p']) < 4)]\n",
    "plt.scatter(r2['r'],\n",
    "            -np.log10(r2['p']),\n",
    "            color=sns.xkcd_rgb['pale red'],\n",
    "            marker='.')\n",
    "\n",
    "texts = []\n",
    "for t, v in r1[-np.log10(r1['p']) >= 4].iterrows():\n",
    "    texts.append(plt.text(v['r'],\n",
    "                          -np.log10(v['p']),\n",
    "                          cd.get(t, t),\n",
    "                          ha='center',\n",
    "                          va='center',\n",
    "                          fontdict={'size': 9}))\n",
    "# adjust_text(texts,\n",
    "#             arrowprops=dict(arrowstyle='-', color='k'),\n",
    "#             force_points=1)\n",
    "\n",
    "plt.ylim(-0.5, 9)\n",
    "\n",
    "plt.xlabel('Pearson\\'s $r$')\n",
    "plt.ylabel('$-log_{10}(pvalue)$')\n",
    "\n",
    "plt.axvline(0,\n",
    "            color='grey',\n",
    "            ls='dashed')\n",
    "plt.axhline(2,\n",
    "            color='grey',\n",
    "            ls='dashed')\n",
    "plt.axhline(4,\n",
    "            color='grey',\n",
    "            ls='dashed')\n",
    "\n",
    "plt.savefig('volcano_3.png',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True)\n",
    "plt.savefig('volcano_3.svg',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clf1, combined1, df_test1 = rndmfrst(p, k, 's-scores')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clf2, combined2, df_test2 = rndmfrst(p, hpi, 's-scores', 'present')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clf3, combined3, df_test3 = rndmfrst(p, nhpi, 's-scores', 'present')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(4, 4))\n",
    "\n",
    "fpr, tpr, _ = metrics.roc_curve(combined1['mouse'],\n",
    "                                clf1.predict_proba(df_test1)[:, 1])\n",
    "plt.plot(fpr, tpr,\n",
    "         'k-',\n",
    "         label='Virulence (%.2f)' % metrics.auc(fpr, tpr))\n",
    "\n",
    "fpr, tpr, _ = metrics.roc_curve(combined2['mouse'],\n",
    "                                clf2.predict_proba(df_test2)[:, 1])\n",
    "plt.plot(fpr, tpr,\n",
    "         '-',\n",
    "         color=sns.xkcd_rgb['dark sky blue'],\n",
    "         label='HPI presence (%.2f)' % metrics.auc(fpr, tpr))\n",
    "\n",
    "# fpr, tpr, _ = metrics.roc_curve(combined3['mouse'],\n",
    "#                                 clf3.predict_proba(df_test3)[:, 1])\n",
    "# plt.plot(fpr, tpr,\n",
    "#          '-',\n",
    "#          color='r',\n",
    "#          label='Other genes (%.2f)' % metrics.auc(fpr, tpr))\n",
    "\n",
    "plt.plot([0, 1],\n",
    "         [0, 1],\n",
    "         '--',\n",
    "         color=sns.xkcd_rgb['grey'],\n",
    "         label='Random predictor')\n",
    "\n",
    "plt.legend(frameon=True,\n",
    "           loc='lower right')\n",
    "\n",
    "plt.xlabel('False positive rate')\n",
    "plt.ylabel('True positive rate')\n",
    "\n",
    "plt.savefig('roc_auc.png',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True)\n",
    "plt.savefig('roc_auc.svg',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(4, 4))\n",
    "\n",
    "prec, rec, _ = metrics.precision_recall_curve(combined1['mouse'],\n",
    "                                              clf1.predict_proba(df_test1)[:, 1])\n",
    "plt.plot(rec, prec,\n",
    "         'k-',\n",
    "         label='Virulence')\n",
    "\n",
    "prec, rec, _ = metrics.precision_recall_curve(combined2['mouse'],\n",
    "                                              clf2.predict_proba(df_test2)[:, 1])\n",
    "plt.plot(rec, prec,\n",
    "         '-',\n",
    "         color=sns.xkcd_rgb['dark sky blue'],\n",
    "         label='HPI presence',\n",
    "         zorder=0)\n",
    "\n",
    "# prec, rec, _ = metrics.precision_recall_curve(combined3['mouse'],\n",
    "#                                               clf3.predict_proba(df_test2)[:, 1])\n",
    "# plt.plot(rec, prec,\n",
    "#          '-',\n",
    "#          color='r',\n",
    "#          label='Other genes',\n",
    "#          zorder=0)\n",
    "\n",
    "plt.axhline(combined1[combined1['mouse'] == 1].shape[0] / combined1.shape[0],\n",
    "            color=sns.xkcd_rgb['grey'],\n",
    "            ls='dashed',\n",
    "            label='Random predictor')\n",
    "\n",
    "# plt.legend(frameon=True,\n",
    "#            loc='lower right')\n",
    "\n",
    "plt.xlabel('Recall')\n",
    "plt.ylabel('Precision')\n",
    "\n",
    "plt.savefig('prec_rec.png',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True)\n",
    "plt.savefig('prec_rec.svg',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fi1 = pd.Series(clf1.feature_importances_,\n",
    "                p['s-scores'].unstack().T.columns)\n",
    "fi1.name = 'Virulence'\n",
    "fi2 = pd.Series(clf2.feature_importances_,\n",
    "                p['s-scores'].unstack().T.columns)\n",
    "fi2.name = 'HPI presence'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fi = fi1.to_frame().join(fi2.to_frame())\n",
    "fi = fi.sort_values('Virulence',\n",
    "                    ascending=False).head(15)\n",
    "fi = fi.stack().reset_index()\n",
    "fi.columns = ['Condition',\n",
    "              'Predictor',\n",
    "              'Feature importance']\n",
    "fi['Condition'] = [cd.get(x, x)\n",
    "                   for x in fi['Condition']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(5, 3.5))\n",
    "\n",
    "sns.barplot(data=fi,\n",
    "            x='Condition',\n",
    "            y='Feature importance',\n",
    "            hue='Predictor',\n",
    "            palette=['k',\n",
    "                     sns.xkcd_rgb['dark sky blue']])\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "plt.savefig('feature_importance.png',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True)\n",
    "plt.savefig('feature_importance.svg',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fi = fi1.to_frame().join(fi2.to_frame())\n",
    "fi = fi.sort_values('Virulence',\n",
    "                    ascending=False)\n",
    "fi.to_csv('feature_importance.tsv',\n",
    "          sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fi1 = pd.Series(clf1.feature_importances_,\n",
    "                p['s-scores'].unstack().T.columns)\n",
    "fi1.name = 'Virulence'\n",
    "fi2 = pd.Series(clf2.feature_importances_,\n",
    "                p['s-scores'].unstack().T.columns)\n",
    "fi2.name = 'HPI presence'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fi = fi1.to_frame().join(fi2.to_frame())\n",
    "fi = fi.sort_values('HPI presence',\n",
    "                    ascending=False).head(15)\n",
    "fi = fi.stack().reset_index()\n",
    "fi.columns = ['Condition',\n",
    "              'Predictor',\n",
    "              'Feature importance']\n",
    "fi['Condition'] = [cd.get(x, x)\n",
    "                   for x in fi['Condition']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(5, 3.5))\n",
    "\n",
    "sns.barplot(data=fi,\n",
    "            x='Condition',\n",
    "            y='Feature importance',\n",
    "            hue='Predictor',\n",
    "            palette=['k',\n",
    "                     sns.xkcd_rgb['dark sky blue']])\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "plt.savefig('feature_importance_2.png',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True)\n",
    "plt.savefig('feature_importance_2.svg',\n",
    "            dpi=300, bbox_inches='tight',\n",
    "            transparent=True);"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
