{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_hits = '../out/associations/summary_cont_lmm_kmer.tsv'\n",
    "names = '../out/associations/associated_ogs.final.tsv'\n",
    "pangenome = '../out/roary/gene_presence_absence.csv'\n",
    "ref_strain = 'IAI39'\n",
    "ref_annotation_dir = '../out/annotations'\n",
    "hpi = '../data/hpi.tsv'\n",
    "outdir = '../out/plots/maps'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir1 = outdir + '1'\n",
    "outdir2 = outdir + '2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_genome = '%s/%s/%s.gbk' % (ref_annotation_dir, ref_strain, ref_strain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib import colors as cl\n",
    "\n",
    "sns.set_style('white')\n",
    "\n",
    "plt.rc('font', size=12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from reportlab.lib import colors\n",
    "from reportlab.lib.units import cm\n",
    "from Bio.Graphics import GenomeDiagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rd = {}\n",
    "roary = pd.read_table(pangenome,\n",
    "                      sep=',',\n",
    "                      low_memory=False)\n",
    "roary.set_index('Gene', inplace=True)\n",
    "# Drop the other info columns\n",
    "roary.drop(list(roary.columns[:13]), axis=1, inplace=True)\n",
    "roary.reset_index(inplace=True)\n",
    "for strain in roary.columns[1:]:\n",
    "    for x, y, in roary.set_index(strain)['Gene'].dropna().to_dict().items():\n",
    "        if str(x) == 'nan':\n",
    "            continue\n",
    "        # be aware of paralogs\n",
    "        for g in x.split('\\t'):\n",
    "            rd[g] = y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "u = pd.read_table(names)\n",
    "n = u.set_index('query')['preferred_og_name'].to_dict()\n",
    "nsize = u.set_index('query')['representative_protein_length']\n",
    "k = pd.read_table(kmer_hits,\n",
    "                  index_col=0)\n",
    "k['normalized_hits'] = (k['length'] / nsize)\n",
    "ogs = set(k[k['normalized_hits'] >= 0.1].index)\n",
    "cmap = plt.get_cmap('viridis')\n",
    "cmap.set_under(sns.xkcd_rgb['light grey'])\n",
    "norm = cl.Normalize(vmin=0.,vmax=k['normalized_hits'].max())\n",
    "khits = k['normalized_hits'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hpic = {x.split()[0]: c\n",
    "        for x, c in zip(open(hpi),\n",
    "                        sns.color_palette('Set2', len(open(hpi).readlines())))}\n",
    "hpic['irp1'] = sns.color_palette('tab20', 2)[0]\n",
    "hpic['irp3'] = sns.color_palette('tab20', 2)[1]\n",
    "ogs = set(hpic.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = SeqIO.read(ref_genome,\n",
    "                 'genbank')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feats = [rd.get(f.qualifiers.get('locus_tag', [''])[0],\n",
    "                f.qualifiers.get('locus_tag', [''])[0])\n",
    "         for f in seq.features\n",
    "         if f.type == 'gene' \n",
    "         and 'locus_tag' in f.qualifiers]\n",
    "idxs = [feats.index(x) for x in ogs\n",
    "        if x in feats]\n",
    "ext_ogs = feats[min(idxs)-10:max(idxs)+10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "img = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gd = GenomeDiagram.Diagram('associated')\n",
    "track = gd.new_track(1,\n",
    "                     name='Reference strain (%s)' % ref_strain,\n",
    "                     greytrack=True)\n",
    "features = track.new_set()\n",
    "\n",
    "pmin = None\n",
    "pmax = None\n",
    "for f in seq.features:\n",
    "    if f.type != 'gene' or 'locus_tag' not in f.qualifiers:\n",
    "        continue\n",
    "    locus = f.qualifiers.get('locus_tag', [''])[0]\n",
    "    og = rd.get(locus, locus)\n",
    "    if og not in ext_ogs:\n",
    "        continue\n",
    "    gmin = int(f.location.start)\n",
    "    gmax = int(f.location.end)\n",
    "    if pmin is None or gmin < pmin:\n",
    "        pmin = gmin\n",
    "    if pmax is None or gmax > pmax:\n",
    "        pmax = gmax\n",
    "    if og == 'pks2':\n",
    "        if gmax - gmin < 2000:\n",
    "            name = 'irp3'\n",
    "            og  = 'irp3'\n",
    "        else:\n",
    "            name = 'irp1'\n",
    "            og = 'irp1'\n",
    "    else:\n",
    "        name = n.get(og, og)\n",
    "    features.add_feature(f,\n",
    "                         label=True,\n",
    "                         name=name,\n",
    "                         color=hpic.get(og, sns.xkcd_rgb['light grey']),\n",
    "                         label_size=5,\n",
    "                         label_angle=90,\n",
    "                         label_position='middle',\n",
    "                         sigil='ARROW',\n",
    "                         arrowshaft_height=0.5)\n",
    "gd.draw(format='linear',\n",
    "        fragments=1,\n",
    "        orientation='landscape',\n",
    "        start=pmin, end=pmax,\n",
    "        pagesize=(((pmax - pmin)/1000)*cm, 3*cm))\n",
    "gd.write(os.path.join(outdir, '%03d_ref_chromosome.svg' % img), 'SVG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for strain in sorted(set(roary.columns[1:]).difference([ref_strain,])):\n",
    "    img += 1\n",
    "    for seq in SeqIO.parse('%s/%s/%s.gbk' % (ref_annotation_dir,\n",
    "                                             strain, strain),\n",
    "                           'genbank'):\n",
    "        feats = [rd.get(f.qualifiers['locus_tag'][0],\n",
    "                        f.qualifiers['locus_tag'][0])\n",
    "                 for f in seq.features\n",
    "                 if f.type == 'gene'\n",
    "                 and 'locus_tag' in f.qualifiers]\n",
    "        if len(set(feats).intersection(ogs)) == 0:\n",
    "            continue\n",
    "        idxs = [feats.index(x) for x in ogs\n",
    "                if x in feats]\n",
    "        ext_feats = feats[min(idxs)-10 if min(idxs)-10 > 10 else 0:max(idxs)+10]\n",
    "        gd = GenomeDiagram.Diagram('test')\n",
    "        track = gd.new_track(1,\n",
    "                             name=strain,\n",
    "                             greytrack=True)\n",
    "        features = track.new_set()\n",
    "\n",
    "        pmin = None\n",
    "        pmax = None\n",
    "        for f in seq.features:\n",
    "            if f.type != 'gene' or 'locus_tag' not in f.qualifiers:\n",
    "                continue\n",
    "            locus = f.qualifiers.get('locus_tag', [''])[0]\n",
    "            og = rd.get(locus, locus)\n",
    "            if og not in ext_feats:\n",
    "                continue\n",
    "            gmin = int(f.location.start)\n",
    "            gmax = int(f.location.end)\n",
    "            if pmin is None or gmin < pmin:\n",
    "                pmin = gmin\n",
    "            if pmax is None or gmax > pmax:\n",
    "                pmax = gmax\n",
    "            if og == 'pks2':\n",
    "                if gmax - gmin < 2000:\n",
    "                    name = 'irp3'\n",
    "                    og  = 'irp3'\n",
    "                else:\n",
    "                    name = 'irp1'\n",
    "                    og = 'irp1'\n",
    "            else:\n",
    "                name = n.get(og, og)\n",
    "            features.add_feature(f,\n",
    "                                 label=True,\n",
    "                                 name=name,\n",
    "                                 color=hpic.get(og, sns.xkcd_rgb['light grey']),\n",
    "                                 label_size=5,\n",
    "                                 label_angle=90,\n",
    "                                 label_position='middle',\n",
    "                                 sigil='ARROW',\n",
    "                                 arrowshaft_height=0.5)\n",
    "        gd.draw(format='linear',\n",
    "                fragments=1,\n",
    "                orientation='landscape',\n",
    "                start=pmin, end=pmax,\n",
    "                pagesize=(((pmax - pmin)/1000)*cm, 3*cm))\n",
    "        gd.write(os.path.join(outdir, '%03d_%s_%s.svg' % (img, strain, seq.id)), 'SVG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = [x for x in os.listdir(outdir)\n",
    "         if 'final' not in x\n",
    "         and x.endswith('.svg')]\n",
    "for img in sorted({x.split('_')[0] for x in files}):\n",
    "    gfiles = [x for x in files\n",
    "              if x.startswith(img)]\n",
    "    os.system('python3 ../src/svg_stack.py %s/%s_* --margin 5px --direction h > %s/%sfinal.svg' % (outdir, img,\n",
    "                                                                                                   outdir, img))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('python3 ../src/svg_stack.py %s/*final.svg --margin 5px --direction v > %s/all.svg' % (outdir,\n",
    "                                                                                                 outdir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ogs = {'iutA', 'iucD', 'iucC'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hpic = {x.split()[0]: c\n",
    "        for x, c in zip(ogs,\n",
    "                        sns.color_palette('Set2', len(ogs)))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = SeqIO.read(ref_genome,\n",
    "                 'genbank')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feats = [rd.get(f.qualifiers.get('locus_tag', [''])[0],\n",
    "                f.qualifiers.get('locus_tag', [''])[0])\n",
    "         for f in seq.features\n",
    "         if f.type == 'gene' \n",
    "         and 'locus_tag' in f.qualifiers]\n",
    "idxs = [feats.index(x) for x in ogs\n",
    "        if x in feats]\n",
    "ext_ogs = feats[min(idxs)-10:max(idxs)+10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "img = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gd = GenomeDiagram.Diagram('associated')\n",
    "track = gd.new_track(1,\n",
    "                     name='Reference strain (%s)' % ref_strain,\n",
    "                     greytrack=True)\n",
    "features = track.new_set()\n",
    "\n",
    "pmin = None\n",
    "pmax = None\n",
    "for f in seq.features:\n",
    "    if f.type != 'gene' or 'locus_tag' not in f.qualifiers:\n",
    "        continue\n",
    "    locus = f.qualifiers.get('locus_tag', [''])[0]\n",
    "    og = rd.get(locus, locus)\n",
    "    if og not in ext_ogs:\n",
    "        continue\n",
    "    gmin = int(f.location.start)\n",
    "    gmax = int(f.location.end)\n",
    "    if pmin is None or gmin < pmin:\n",
    "        pmin = gmin\n",
    "    if pmax is None or gmax > pmax:\n",
    "        pmax = gmax\n",
    "    if og == 'pks2':\n",
    "        if gmax - gmin < 2000:\n",
    "            name = 'irp3'\n",
    "            og  = 'irp3'\n",
    "        else:\n",
    "            name = 'irp1'\n",
    "            og = 'irp1'\n",
    "    else:\n",
    "        name = n.get(og, og)\n",
    "    features.add_feature(f,\n",
    "                         label=True,\n",
    "                         name=name,\n",
    "                         color=hpic.get(og, sns.xkcd_rgb['light grey']),\n",
    "                         label_size=5,\n",
    "                         label_angle=90,\n",
    "                         label_position='middle',\n",
    "                         sigil='ARROW',\n",
    "                         arrowshaft_height=0.5)\n",
    "gd.draw(format='linear',\n",
    "        fragments=1,\n",
    "        orientation='landscape',\n",
    "        start=pmin, end=pmax,\n",
    "        pagesize=(((pmax - pmin)/1000)*cm, 3*cm))\n",
    "gd.write(os.path.join(outdir1, '%03d_ref_chromosome.svg' % img), 'SVG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for strain in sorted(set(roary.columns[1:]).difference([ref_strain,])):\n",
    "    img += 1\n",
    "    for seq in SeqIO.parse('%s/%s/%s.gbk' % (ref_annotation_dir,\n",
    "                                             strain, strain),\n",
    "                           'genbank'):\n",
    "        feats = [rd.get(f.qualifiers['locus_tag'][0],\n",
    "                        f.qualifiers['locus_tag'][0])\n",
    "                 for f in seq.features\n",
    "                 if f.type == 'gene'\n",
    "                 and 'locus_tag' in f.qualifiers]\n",
    "        if len(set(feats).intersection(ogs)) == 0:\n",
    "            continue\n",
    "        idxs = [feats.index(x) for x in ogs\n",
    "                if x in feats]\n",
    "        ext_feats = feats[min(idxs)-10 if min(idxs)-10 > 10 else 0:max(idxs)+10]\n",
    "        gd = GenomeDiagram.Diagram('test')\n",
    "        track = gd.new_track(1,\n",
    "                             name=strain,\n",
    "                             greytrack=True)\n",
    "        features = track.new_set()\n",
    "\n",
    "        pmin = None\n",
    "        pmax = None\n",
    "        for f in seq.features:\n",
    "            if f.type != 'gene' or 'locus_tag' not in f.qualifiers:\n",
    "                continue\n",
    "            locus = f.qualifiers.get('locus_tag', [''])[0]\n",
    "            og = rd.get(locus, locus)\n",
    "            if og not in ext_feats:\n",
    "                continue\n",
    "            gmin = int(f.location.start)\n",
    "            gmax = int(f.location.end)\n",
    "            if pmin is None or gmin < pmin:\n",
    "                pmin = gmin\n",
    "            if pmax is None or gmax > pmax:\n",
    "                pmax = gmax\n",
    "            if og == 'pks2':\n",
    "                if gmax - gmin < 2000:\n",
    "                    name = 'irp3'\n",
    "                    og  = 'irp3'\n",
    "                else:\n",
    "                    name = 'irp1'\n",
    "                    og = 'irp1'\n",
    "            else:\n",
    "                name = n.get(og, og)\n",
    "            features.add_feature(f,\n",
    "                                 label=True,\n",
    "                                 name=name,\n",
    "                                 color=hpic.get(og, sns.xkcd_rgb['light grey']),\n",
    "                                 label_size=5,\n",
    "                                 label_angle=90,\n",
    "                                 label_position='middle',\n",
    "                                 sigil='ARROW',\n",
    "                                 arrowshaft_height=0.5)\n",
    "        gd.draw(format='linear',\n",
    "                fragments=1,\n",
    "                orientation='landscape',\n",
    "                start=pmin, end=pmax,\n",
    "                pagesize=(((pmax - pmin)/1000)*cm, 3*cm))\n",
    "        gd.write(os.path.join(outdir1, '%03d_%s_%s.svg' % (img, strain, seq.id)), 'SVG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = [x for x in os.listdir(outdir1)\n",
    "         if 'final' not in x\n",
    "         and x.endswith('.svg')]\n",
    "for img in sorted({x.split('_')[0] for x in files}):\n",
    "    gfiles = [x for x in files\n",
    "              if x.startswith(img)]\n",
    "    os.system('python3 ../src/svg_stack.py %s/%s_* --margin 5px --direction h > %s/%sfinal.svg' % (outdir1, img,\n",
    "                                                                                                   outdir1, img))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('python3 ../src/svg_stack.py %s/*final.svg --margin 5px --direction v > %s/all.svg' % (outdir1,\n",
    "                                                                                                 outdir1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ogs = {'mntB', 'mntB_2', 'btuD_2', 'hpf'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hpic = {x.split()[0]: c\n",
    "        for x, c in zip(ogs,\n",
    "                        sns.color_palette('Set2', len(ogs)))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = SeqIO.read(ref_genome,\n",
    "                 'genbank')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feats = [rd.get(f.qualifiers.get('locus_tag', [''])[0],\n",
    "                f.qualifiers.get('locus_tag', [''])[0])\n",
    "         for f in seq.features\n",
    "         if f.type == 'gene' \n",
    "         and 'locus_tag' in f.qualifiers]\n",
    "idxs = [feats.index(x) for x in ogs\n",
    "        if x in feats]\n",
    "ext_ogs = feats[min(idxs)-10:max(idxs)+10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "img = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gd = GenomeDiagram.Diagram('associated')\n",
    "track = gd.new_track(1,\n",
    "                     name='Reference strain (%s)' % ref_strain,\n",
    "                     greytrack=True)\n",
    "features = track.new_set()\n",
    "\n",
    "pmin = None\n",
    "pmax = None\n",
    "for f in seq.features:\n",
    "    if f.type != 'gene' or 'locus_tag' not in f.qualifiers:\n",
    "        continue\n",
    "    locus = f.qualifiers.get('locus_tag', [''])[0]\n",
    "    og = rd.get(locus, locus)\n",
    "    if og not in ext_ogs:\n",
    "        continue\n",
    "    gmin = int(f.location.start)\n",
    "    gmax = int(f.location.end)\n",
    "    if pmin is None or gmin < pmin:\n",
    "        pmin = gmin\n",
    "    if pmax is None or gmax > pmax:\n",
    "        pmax = gmax\n",
    "    if og == 'pks2':\n",
    "        if gmax - gmin < 2000:\n",
    "            name = 'irp3'\n",
    "            og  = 'irp3'\n",
    "        else:\n",
    "            name = 'irp1'\n",
    "            og = 'irp1'\n",
    "    else:\n",
    "        name = n.get(og, og)\n",
    "    features.add_feature(f,\n",
    "                         label=True,\n",
    "                         name=name,\n",
    "                         color=hpic.get(og, sns.xkcd_rgb['light grey']),\n",
    "                         label_size=5,\n",
    "                         label_angle=90,\n",
    "                         label_position='middle',\n",
    "                         sigil='ARROW',\n",
    "                         arrowshaft_height=0.5)\n",
    "gd.draw(format='linear',\n",
    "        fragments=1,\n",
    "        orientation='landscape',\n",
    "        start=pmin, end=pmax,\n",
    "        pagesize=(((pmax - pmin)/1000)*cm, 3*cm))\n",
    "gd.write(os.path.join(outdir2, '%03d_ref_chromosome.svg' % img), 'SVG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for strain in sorted(set(roary.columns[1:]).difference([ref_strain,])):\n",
    "    img += 1\n",
    "    for seq in SeqIO.parse('%s/%s/%s.gbk' % (ref_annotation_dir,\n",
    "                                             strain, strain),\n",
    "                           'genbank'):\n",
    "        feats = [rd.get(f.qualifiers['locus_tag'][0],\n",
    "                        f.qualifiers['locus_tag'][0])\n",
    "                 for f in seq.features\n",
    "                 if f.type == 'gene'\n",
    "                 and 'locus_tag' in f.qualifiers]\n",
    "        if len(set(feats).intersection(ogs)) == 0:\n",
    "            continue\n",
    "        idxs = [feats.index(x) for x in ogs\n",
    "                if x in feats]\n",
    "        ext_feats = feats[min(idxs)-10 if min(idxs)-10 > 10 else 0:max(idxs)+10]\n",
    "        gd = GenomeDiagram.Diagram('test')\n",
    "        track = gd.new_track(1,\n",
    "                             name=strain,\n",
    "                             greytrack=True)\n",
    "        features = track.new_set()\n",
    "\n",
    "        pmin = None\n",
    "        pmax = None\n",
    "        for f in seq.features:\n",
    "            if f.type != 'gene' or 'locus_tag' not in f.qualifiers:\n",
    "                continue\n",
    "            locus = f.qualifiers.get('locus_tag', [''])[0]\n",
    "            og = rd.get(locus, locus)\n",
    "            if og not in ext_feats:\n",
    "                continue\n",
    "            gmin = int(f.location.start)\n",
    "            gmax = int(f.location.end)\n",
    "            if pmin is None or gmin < pmin:\n",
    "                pmin = gmin\n",
    "            if pmax is None or gmax > pmax:\n",
    "                pmax = gmax\n",
    "            if og == 'pks2':\n",
    "                if gmax - gmin < 2000:\n",
    "                    name = 'irp3'\n",
    "                    og  = 'irp3'\n",
    "                else:\n",
    "                    name = 'irp1'\n",
    "                    og = 'irp1'\n",
    "            else:\n",
    "                name = n.get(og, og)\n",
    "            features.add_feature(f,\n",
    "                                 label=True,\n",
    "                                 name=name,\n",
    "                                 color=hpic.get(og, sns.xkcd_rgb['light grey']),\n",
    "                                 label_size=5,\n",
    "                                 label_angle=90,\n",
    "                                 label_position='middle',\n",
    "                                 sigil='ARROW',\n",
    "                                 arrowshaft_height=0.5)\n",
    "        gd.draw(format='linear',\n",
    "                fragments=1,\n",
    "                orientation='landscape',\n",
    "                start=pmin, end=pmax,\n",
    "                pagesize=(((pmax - pmin)/1000)*cm, 3*cm))\n",
    "        gd.write(os.path.join(outdir2, '%03d_%s_%s.svg' % (img, strain, seq.id)), 'SVG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = [x for x in os.listdir(outdir2)\n",
    "         if 'final' not in x\n",
    "         and x.endswith('.svg')]\n",
    "for img in sorted({x.split('_')[0] for x in files}):\n",
    "    gfiles = [x for x in files\n",
    "              if x.startswith(img)]\n",
    "    os.system('python3 ../src/svg_stack.py %s/%s_* --margin 5px --direction h > %s/%sfinal.svg' % (outdir2, img,\n",
    "                                                                                                   outdir2, img))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('python3 ../src/svg_stack.py %s/*final.svg --margin 5px --direction v > %s/all.svg' % (outdir2,\n",
    "                                                                                                 outdir2))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
